In [1]:
!conda install -c conda-forge scikit-surprise

Solving environment: done

## Package Plan ##

  environment location: /opt/anaconda3/envs/reccomendations

  added / updated specs:
    - scikit-surprise


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    openssl-1.1.1h             |       haf1e3a3_0         1.9 MB  conda-forge
    scikit-surprise-1.1.1      |   py37h57c32b8_1         577 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         2.4 MB

The following packages will be UPDATED:

  certifi               pkgs/main::certifi-2020.6.20-py37_0 --> conda-forge::certifi-2020.6.20-py37h2987424_2
  scikit-surprise                      1.1.1-py37h5be27a9_0 --> 1.1.1-py37h57c32b8_1

The following packages will be SUPERSEDED by a higher-priority channel:

  ca-certificates    pkgs/main::ca-certificates-2020.10.14~ --> conda-forge::ca-certificates-2020.6.20

In [3]:
from surprise import NormalPredictor, BaselineOnly, KNNBasic, KNNWithMeans,  KNNBaseline, SVD, SVDpp
from surprise import Dataset
from surprise.model_selection import cross_validate

# Load the movielens-100k dataset (download it if needed).
data = Dataset.load_builtin('ml-100k')

# Use the famous SVD algorithm.

norm_pred = NormalPredictor()
baseline_only = BaselineOnly()
knn_basic = KNNBasic()
knn_with_means = KNNWithMeans()
knn_baseline = KNNBaseline()
svd = SVD()
svd_pp = SVDpp()
# Run 5-fold cross-validation and print results.
cross_validate(norm_pred, data, measures=['RMSE', 'MAE', 'FCP'], cv=5, verbose=True)
cross_validate(baseline_only, data, measures=['RMSE', 'MAE', 'FCP'], cv=5, verbose=True)
cross_validate(knn_basic, data, measures=['RMSE', 'MAE', 'FCP'], cv=5, verbose=True)
cross_validate(knn_with_means, data, measures=['RMSE', 'MAE', 'FCP'], cv=5, verbose=True)
cross_validate(knn_baseline, data, measures=['RMSE', 'MAE', 'FCP'], cv=5, verbose=True)
cross_validate(svd, data, measures=['RMSE', 'MAE', 'FCP'], cv=5, verbose=True)
cross_validate(svd_pp, data, measures=['RMSE', 'MAE', 'FCP'], cv=5, verbose=True)


Evaluating RMSE, MAE, FCP of algorithm NormalPredictor on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    1.5149  1.5149  1.5208  1.5208  1.5147  1.5172  0.0029  
MAE (testset)     1.2202  1.2140  1.2186  1.2223  1.2188  1.2188  0.0027  
FCP (testset)     0.5034  0.4975  0.4973  0.4986  0.4861  0.4966  0.0057  
Fit time          0.09    0.12    0.12    0.11    0.12    0.11    0.01    
Test time         0.10    0.10    0.10    0.10    0.19    0.12    0.03    
Estimating biases using als...
Estimating biases using als...
Estimating biases using als...
Estimating biases using als...
Estimating biases using als...
Evaluating RMSE, MAE, FCP of algorithm BaselineOnly on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9457  0.9513  0.9435  0.9427  0.9379  0.9442  0.0044  
MAE (testset)     0.7485  0.7531  0.7498  0.7489  0.7422  0.7485  0.0035  
FCP (testset)     0.7030  0.690

{&#39;test_rmse&#39;: array([0.91785677, 0.92808664, 0.92061762, 0.92556979, 0.91288353]),
 &#39;test_mae&#39;: array([0.72217461, 0.72494952, 0.72136172, 0.72863724, 0.7161946 ]),
 &#39;test_fcp&#39;: array([0.711361  , 0.70527626, 0.70542808, 0.71121048, 0.71359846]),
 &#39;fit_time&#39;: (139.5071828365326,
  142.66110515594482,
  144.75219988822937,
  141.42762994766235,
  135.42054295539856),
 &#39;test_time&#39;: (2.5089988708496094,
  2.793266773223877,
  2.7438762187957764,
  2.799530029296875,
  2.6305768489837646)}

---
Normal predictor
---

- RMSE (testset)    1.5149  1.5149  1.5208  1.5208  1.5147  1.5172  0.0029  
- MAE (testset)     1.2202  1.2140  1.2186  1.2223  1.2188  1.2188  0.0027  
- FCP (testset)     0.5034  0.4975  0.4973  0.4986  0.4861  0.4966  0.0057  
---
Baseline only
---

- RMSE (testset)    0.9457  0.9513  0.9435  0.9427  0.9379  0.9442  0.0044  
- MAE (testset)     0.7485  0.7531  0.7498  0.7489  0.7422  0.7485  0.0035  
- FCP (testset)     0.7030  0.6907  0.6995  0.6929  0.6969  0.6966  0.0044  
---
KNN Basic
---

- RMSE (testset)    0.9760  0.9785  0.9727  0.9804  0.9868  0.9789  0.0047  
- MAE (testset)     0.7689  0.7720  0.7700  0.7739  0.7801  0.7730  0.0040  
- FCP (testset)     0.7129  0.7066  0.7071  0.7164  0.7042  0.7094  0.0045  
---
KNN with Means
---

- RMSE (testset)    0.9502  0.9461  0.9596  0.9510  0.9410  0.9496  0.0061  
- MAE (testset)     0.7459  0.7455  0.7552  0.7524  0.7412  0.7480  0.0050  
- FCP (testset)     0.7015  0.7002  0.7063  0.7062  0.7049  0.7038  0.0025 
---
KNN Baseline
---

- RMSE (testset)    0.9272  0.9347  0.9327  0.9367  0.9211  0.9305  0.0057  
- MAE (testset)     0.7293  0.7355  0.7350  0.7381  0.7269  0.7329  0.0042  
- FCP (testset)     0.7080  0.7070  0.7039  0.7034  0.7177  0.7080  0.0052 
---
SVD
---

- RMSE (testset)    0.9408  0.9338  0.9317  0.9400  0.9310  0.9354  0.0041  
- MAE (testset)     0.7414  0.7362  0.7348  0.7411  0.7326  0.7372  0.0035  
- FCP (testset)     0.6978  0.6978  0.7033  0.6978  0.7028  0.6999  0.0026
---
SVDpp
---

- RMSE (testset)    0.9179  0.9281  0.9206  0.9256  0.9129  0.9210  0.0054  
- MAE (testset)     0.7222  0.7249  0.7214  0.7286  0.7162  0.7227  0.0041  
- FCP (testset)     0.7114  0.7053  0.7054  0.7112  0.7136  0.7094  0.0034    
---

In [1]:
from lightfm_dataset import fetch_movielens

In [2]:
from sklearn.metrics.pairwise import cosine_similarity
import numpy as np

def compute_personalization(predictions):
    cosine_similarity_matrix = cosine_similarity(predictions)
    cosine_similarity_matrix_mean = np.triu(cosine_similarity_matrix).mean()
    return 1 - cosine_similarity_matrix_mean



In [3]:
from lightfm.lightfm import LightFM

model = LightFM(no_components=10)
dataset = fetch_movielens(test_part=0.2)

In [4]:
train = dataset['train']
model.fit(train)

In [5]:
from lightfm.evaluation import precision_at_k, recall_at_k
k_val = 3
prec = precision_at_k(model, dataset['test'], k=k_val).mean()
rec = recall_at_k(model, dataset['test'], k=k_val ).mean()
print(f"Precision at {k_val}: {prec}")
print(f"Recall at {k_val}: {rec}")

Precision at 3: 0.45484402775764465
Recall at 3: 0.020012179507967427


In [6]:
from surprise import NormalPredictor, BaselineOnly, KNNBasic, KNNWithMeans,  KNNBaseline, SVD, SVDpp
from surprise_dataset import get_surprise_ml_100k
from tqdm import tqdm
normal_pred = NormalPredictor()
data = get_surprise_ml_100k(0.2)
train = data[0]
train = get_surprise_ml_100k(0.000001)[0]
test = list(map(lambda x: (int(x[0]), int(x[1])),data[1]))
test_items = sorted(list(set(map(lambda x: x[1], test))))
test_users = sorted(list(set(map(lambda x: x[0], test))))


normal_pred.fit(train)

predictions = { 'normal_pred': {}, 'baseline_only': {},'knn_basic': {},'knn_with_means':{}, 'knn_baseline': {}, 'svd':{}, 'svd_pp':{} }

for user in tqdm(test_users):
    preds = []
    for item in test_items:
        pred = normal_pred.estimate(user, item)
        preds.append(pred) 
    predictions['normal_pred'][user] = preds


100%|██████████| 942/942 [00:03<00:00, 240.57it/s]


In [7]:
 import pickle
 pickle.dump( predictions, open( "predictions.p", "wb" ) )

In [8]:
baseline_only = BaselineOnly()

baseline_only.fit(train)
for user in tqdm(test_users):
    preds = []
    for item in test_items:
        pred = baseline_only.estimate(user, item)
        preds.append(pred) 
    predictions['baseline_only'][user] = preds

knn_basic = KNNBasic()
knn_basic.fit(train)
#for inst in test:
    #print(f"{inst}")
    #pred = knn_basic.estimate(str(inst[0]), str(inst[1]))
        
    #predictions['knn_basic'].append((inst[0], inst[1], pred))

knn_with_means = KNNWithMeans()
knn_with_means.fit(train)
#for inst in test:
#    pred = knn_with_means.estimate(inst[0], inst[1])
#    predictions['knn_with_means'].append((inst[0], inst[1], pred))

knn_baseline = KNNBaseline()
knn_baseline.fit(train)
#for inst in test:
#    pred = knn_baseline.estimate(inst[0], inst[1])
#    predictions['knn_baseline'].append((inst[0], inst[1], pred))

svd = SVD()
svd.fit(train)
for user in tqdm(test_users):
    preds = []
    for item in test_items:
        pred = svd.estimate(user, item)
        preds.append(pred) 
    predictions['svd'][user] = preds

svd_pp = SVDpp()
svd_pp.fit(train)
for user in tqdm(test_users):
    preds = []
    for item in test_items:
        pred = svd_pp.estimate(user, item)
        preds.append(pred) 
    predictions['svd_pp'][user] = preds

Estimating biases using als...
100%|██████████| 942/942 [00:01<00:00, 575.84it/s]
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Estimating biases using als...
Computing the msd similarity matrix...
Done computing similarity matrix.
100%|██████████| 942/942 [01:41<00:00,  9.27it/s]


In [11]:

personalization([[1,2,3],[0,1,5], [6,1,7]])

0.6666666666666665

In [12]:
from metrics import personalization, mapk, mark

In [16]:
#NormalPredictor
one_hot_encoded_preds = list(predictions['normal_pred'].values())
for i in range(len(one_hot_encoded_preds)):
    one_hot_encoded_preds[i] = list(map(lambda x: 1 if x>=4 else 0, one_hot_encoded_preds[i]))
his = personalization(predicted=one_hot_encoded_preds)
from metrics import mark
print(his)
actual = {key:{} for key in test_users}
validation_data = list(map(lambda x: (int(x[0]), int(x[1]), int(x[2])),data[1]))
for inst in validation_data:
    actual[inst[0]][inst[1]] = inst[2] 



data_recall = [[],[]]

for user in actual.keys():
    y = []
    for i in (list(actual[user].keys())):
        if actual[user][i]>=4:
            y.append(i)
        else:
            y.append(0)
    data_recall[0].append(y)
    need_to_predict = actual[user].keys()
    pred = []
    for item in need_to_predict:
        if  normal_pred.estimate(user, item)>=4:
            pred.append( item)
        else:
            pred.append(0)
    data_recall[1].append(pred)
    

mark_v = mark(data_recall[0], data_recall[1])
print(mark_v)
mapk_v = mapk(data_recall[0], data_recall[1])
print(mapk_v)

1.1102230246251565e-16
0.17763832065483076
0.24421645740154993


In [17]:
#BaselineOnly
one_hot_encoded_preds = list(predictions['baseline_only'].values())
for i in range(len(one_hot_encoded_preds)):
    one_hot_encoded_preds[i] = list(map(lambda x: 1 if x>=4 else 0, one_hot_encoded_preds[i]))
p = personalization(one_hot_encoded_preds)
print (p)

data_recall = [[],[]]
for user in actual.keys():
    y = []
    for i in (list(actual[user].keys())):
        if actual[user][i]>=4:
            y.append(i)
    data_recall[0].append(y)
    need_to_predict = actual[user].keys()
    pred = []
    for item in need_to_predict:
        if  baseline_only.estimate(user, item)>=4:
            pred.append( item)
    data_recall[1].append(pred)

mark_v = mark(data_recall[0], data_recall[1])
print(mark_v)
mapk_v = mapk(data_recall[0], data_recall[1])
print(mapk_v)

0.03605563494241382
0.11995842304494037
0.1696995240850134


In [18]:
#SVD 
one_hot_encoded_preds = list(predictions['svd'].values())
for i in range(len(one_hot_encoded_preds)):
    one_hot_encoded_preds[i] = list(map(lambda x: 1 if x>=4 else 0, one_hot_encoded_preds[i]))
p = personalization(one_hot_encoded_preds)
print (p)

data_recall = [[],[]]
for user in actual.keys():
    y = []
    for i in (list(actual[user].keys())):
        if actual[user][i]>=4:
            y.append(i)
    data_recall[0].append(y)
    need_to_predict = actual[user].keys()
    pred = []
    for item in need_to_predict:
        if  svd.estimate(user, item)>=4:
            pred.append( item)
    data_recall[1].append(pred)

mark_v = mark(data_recall[0], data_recall[1])
print(mark_v)
mapk_v = mapk(data_recall[0], data_recall[1])
print(mapk_v)

0.010391767952152575
0.13298764189593376
0.18748190243896568


In [19]:
#SVDPP 
one_hot_encoded_preds = list(predictions['svd_pp'].values())
for i in range(len(one_hot_encoded_preds)):
    one_hot_encoded_preds[i] = list(map(lambda x: 1 if x>=4 else 0, one_hot_encoded_preds[i]))
p = personalization(one_hot_encoded_preds)
print (p)

data_recall = [[],[]]
for user in actual.keys():
    y = []
    for i in (list(actual[user].keys())):
        if actual[user][i]>=4:
            y.append(i)
    data_recall[0].append(y)
    need_to_predict = actual[user].keys()
    pred = []
    for item in need_to_predict:
        if  svd_pp.estimate(user, item)>=4:
            pred.append( item)
    data_recall[1].append(pred)

mark_v = mark(data_recall[0], data_recall[1])
print(mark_v)
mapk_v = mapk(data_recall[0], data_recall[1])
print(mapk_v)


0.02097848481024378
0.1344802957822535
0.18609252692977796
